samples/nrf9160/serial_lte_modem/src/tcpip/th_tcpip_service.c

```diff
@@ -0,0 +1,333 @@
+ */
+int do_tcp_disconnect(void)
+{
+	int ret;
+
+	if (!client.connected) {
+		LOG_ERR("Not connected yet");
+		return -EINVAL;
+	}
+
+	ret = close(client.sock);
+	if (ret < 0) {
+		LOG_ERR("close() failed: %d", -errno);
+		return -errno;
+	}
+
+	client.connected = false;
+	LOG_INF("TCP disconnected");
+	return 0;
+}
+
+/**@brief Handles send requests on TCP socket transport.
+ */
+int do_tcp_send(const u8_t *data, u32_t datalen)
+{
+	u32_t offset = 0;
+	int ret;
+
+	if (!client.connected) {
+		LOG_ERR("Not connected yet");
+		return -EINVAL;
+	}
+
+	LOG_HEXDUMP_DBG(data, datalen, "TX");
+
+	while (offset < datalen) {
+		ret = send(client.sock, data + offset,
+			   datalen - offset, 0);
+		if (ret < 0) {
+			LOG_ERR("send() failed: %d", -errno);
+			return -errno;
+		}
+
+		offset += ret;
+	}
+
+	LOG_INF("TCP sent");
+	return 0;
+}
+
+/**@brief Handles receive requests on TCP socket transport.
+ */
+int do_tcp_receive(void)
+{
+	int ret;
+
+	if (!client.connected) {
+		LOG_ERR("Not connected yet");
+		return -EINVAL;
+	}
+
+	while (true) {
+		ret = recv(client.sock, rx_buffer,
+			TCPIP_BUFFER_SIZE,
+			0);
+		if (ret < 0) {
+			LOG_WRN("recv() error: %d", -errno);
+			break;
+		}
+		if (ret > 0) {
+			LOG_HEXDUMP_DBG(rx_buffer, ret, "RX");
+			inter_connect_notify(NOT_TYPE_TCP, rx_buffer, ret);
+		}
+	}
+
+	LOG_INF("TCP received");
+	return 0;
+}
+
+/**@brief init UDP socket transport.
+ */
+static int udp_init(const u8_t *param, u8_t length)
+{
+	u16_t port;
+	bool address_use_ip;
+	u8_t address_length;
+	char address[128];
+	int ret;
+
+	memcpy(&port, param, 2);
+	address_use_ip = *(param+2) & ADDRESS_TYPE_MASK;
+	address_length = *(param+2) & ADDRESS_LENGTH_MASK;
+	memcpy(address, param + 3, address_length);
+
+	if (address_use_ip) {
+		ret = util_parse_host_by_ipv4(address, port,
+					(void *)&client.remote);
+	} else {
+		ret = util_parse_host_by_name(address, port,
+					(void *)&client.remote);
+	}
+	if (ret) {
+		LOG_ERR("Parse failed: %d", ret);
+		return ret;
+	}
+
+	client.sock = socket(client.remote.sa_family, SOCK_DGRAM, IPPROTO_UDP);
+	if (client.sock < 0) {
+		LOG_ERR("socket() failed: %d", -errno);
+		return -errno;
+	}
+
+	LOG_INF("UDP initialized");
+	return 0;
+}
+
+/**@brief Handles sendto requests on UDP socket transport.
+ */
+int do_udp_sendto(const u8_t *param, u8_t length)
+{
+	int ret;
+	u8_t *data;
+	u8_t datalen;
+	u8_t address_length = *(param+2) & ADDRESS_LENGTH_MASK;
+	u32_t offset = 0;
+
+	ret = udp_init(param, length);
+	if (ret) {
+		return ret;
+	}
+
+	address_length = *(param+2) & ADDRESS_LENGTH_MASK;
+	data = (u8_t *)param + address_length + 3;
+	datalen = length - address_length - 3;
+
+	LOG_HEXDUMP_DBG(data, datalen, "TX");
+
+	while (offset < datalen) {
+		ret = sendto(client.sock, data + offset,
+			   datalen - offset, 0,
+			   (struct sockaddr *)&client.remote,
+			   sizeof(struct sockaddr));
+		if (ret < 0) {
+			LOG_ERR("sendto() failed: %d", -errno);
+			return -errno;
```

Comment:
We should close the socket here, otherwise, we leak a socket. Or `break` instead of `return`.